Guides/BSON.md

```diff
@@ -0,0 +1,429 @@
+
+For example, here is an function written using raw documents:
+```swift
+let person = [
+    "name": "Bob",
+    "occupation": "Software Engineer"
+    "projects": [
+        ["id": 1, title: "Server Side Swift Application"],
+        ["id": 76, title: "Write documentation"],
+    ]
+]
+
+func prettyPrint(doc: Document) throws {
+    guard let name = doc["name"]?.stringValue else {
+        throw argumentError(message: "missing name")
+    }
+    print("Name: \(name)")
+    guard let occupation = doc["occupation"]?.stringValue else {
+        throw argumentError(message: "missing occupation")
+    }
+    print("Occupation: \(occupation)")
+    guard let projects = doc["projects"]?.arrayValue.compactMap { $0.documentValue } else {
+        throw argumentError(message: "missing projects")
+    }
+    print("Projects:")
+    for project in projects {
+        guard let title = project["title"] else {
+            throw argumentError(message: "missing title")
+        }
+        print(title)
+    }
+}
+```
+Due to the flexible nature of `Document`, a number of checks have to be put into the body of the function. This clutters the actual function's logic and requires a lot of boilerplate code. Now, consider the following function which does the same thing but is written leveraging `Codable`:
+```swift
+struct Project: Codable {
+    let id: BSON
+    let title: String
+}
+
+struct Person: Codable {
+    let name: String
+    let occupation: String
+    let projects: [Project]
+}
+
+func prettyPrint(doc: Document) throws {
+    let person = try BSONDecoder().decode(Person.self, from: doc)
+    print("Name: \(person.name)")
+    print("Occupation: \(person.occupation)")
+    print("Projects:")
+    for project in person.projects {
+        print(project.title)
+    }
+}
+```
+In this version, the definition of the data type and the logic of the function are defined completely separately, and it leads to far more readable and concise versions of both. 
+
+### `Codable` in MongoSwift
+There are a number of ways for users to leverage `Codable` via driver's API. One such example is through `MongoCollection<T>`. By default, `MongoDatabase::collection` returns a `MongoCollection<Document>`. Any `find` or `aggregate` method invocation on that returned collection would then return a `MongoCursor<Document>`, which when iterated returns a `Document?`:
```

Comment:
`MongoDatabase::collection`
nit; maybe just `MongoDatabase.collection`? `::` not very Swifty

Reply:
fixed